File: 206-ReverseLinkedList/206-ReverseLinkedList.py
# Last updated: 8/13/2025, 8:21:18 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head

        while curr is not None : 
            curr_next = curr.next
            curr.next = prev
            prev = curr
            curr = curr_next
        
        return prev

File: 206-ReverseLinkedList/206-ReverseLinkedList.java
// Last updated: 8/13/2025, 8:21:24 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head==null || head.next==null){
            return head;
        }

        ListNode newHead = reverseList(head.next);
        ListNode front = head.next;
        front.next = head;
        head.next = null;

        return newHead;
        
    }
}